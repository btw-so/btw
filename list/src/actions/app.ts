import { createAction } from '@reduxjs/toolkit';

import { ActionTypes } from 'literals';

import { SetAppOOptions } from 'types';

export const setAppOptions = createAction<SetAppOOptions>(ActionTypes.SET_APP_OPTIONS);
