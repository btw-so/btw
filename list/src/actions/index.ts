export * from './alerts';
export * from './app';
export * from './github';
export * from './user';
export * from './list';
export * from './notes';