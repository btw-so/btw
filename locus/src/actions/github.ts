import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from 'modules/helpers';

import { ActionTypes } from 'literals';

export interface GetReposSuccessMeta {
  cached: boolean;
  query: string;
  updatedAt: number;
}

export const getRepos = createAction<string>(ActionTypes.GITHUB_GET_REPOS_REQUEST);
export const getReposSuccess = createAction(
  ActionTypes.GITHUB_GET_REPOS_SUCCESS,
  (payload: Record<string, any>[], meta: GetReposSuccessMeta) => actionPayload(payload, meta),
);
export const getReposFailure = createAction(
  ActionTypes.GITHUB_GET_REPOS_FAILURE,
  (payload: string, query: string) => actionPayload(payload, { query }),
);
