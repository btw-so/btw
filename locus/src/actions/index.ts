export * from './alerts';
export * from './app';
export * from './github';
export * from './user';
export * from './list';
export * from './notes';
export * from './files';
export * from './intelligence';