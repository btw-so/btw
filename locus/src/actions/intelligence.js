
import { createAction } from "@reduxjs/toolkit";

import { ActionTypes } from "literals";

export const saveIntelligenceApiKeys = createAction(
  ActionTypes.SAVE_INTELLIGENCE_API_KEYS,
  (apiKeys) => ({
    payload: { apiKeys },
  })
);

export const savePreferredTabs = createAction(
  ActionTypes.SAVE_PREFERRED_TABS,
  (preferredTabs) => ({
    payload: { preferredTabs },
  })
);
