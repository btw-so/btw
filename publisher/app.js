var express = require('express')
var path = require('path')
var cookieParser = require('cookie-parser')
var logger = require('morgan')
var exphbs = require('express-handlebars')
var hbsLayout = require('handlebars-layout')

var cron = require('node-cron')

var indexRouter = require('./routes/index')
var internalRouter = require('./routes/internal')
var { genSitemap } = require('./routes/sitemap')

var { cacheUsers, cacheNotes } = require('./logic/notes')

var app = express()
var minifyHTML = require('express-minify-html')

var hbs = exphbs.create({
  extname: '.hbs',
  defaultLayout: 'base',
  layoutsDir: path.join(__dirname, '/views/layouts'),
  partialsDir: path.join(__dirname, '/views/partials'),
  helpers: require('handlebars-helpers')(), // make sure to call the returned function
})

hbs.handlebars.registerHelper(hbsLayout(hbs.handlebars))

// register new function
hbs.handlebars.registerHelper('slicer', (str, s, e = null) => str.slice(s, e))
hbs.handlebars.registerHelper('ifDev', (options) => {
  if (process.env.NODE_ENV !== 'production') {
    return options.fn(this)
  } else {
    return options.inverse(this)
  }
})
hbs.handlebars.registerHelper('ifIsNthItem', function (options) {
  var index = options.data.index + 1,
    nth = options.hash.nth,
    val = options.hash.val

  if (index % nth === val || 0) {
    return options.fn(this)
  } else {
    return options.inverse(this)
  }
})
hbs.handlebars.registerHelper('dateddmmyyy', function (str) {
  const intervals = [
    { label: 'year', seconds: 31536000 },
    { label: 'month', seconds: 2592000 },
    { label: 'day', seconds: 86400 },
    { label: 'hour', seconds: 3600 },
    { label: 'minute', seconds: 60 },
    { label: 'second', seconds: 1 },
  ]

  function timeSince(date) {
    const seconds = Math.floor((Date.now() - date.getTime()) / 1000)
    const interval = intervals.find((i) => i.seconds < seconds)
    const count = Math.floor(seconds / interval.seconds)
    return `${count} ${interval.label}${count !== 1 ? 's' : ''} ago`
  }

  return timeSince(new Date(str))
})

// Minifier setup
app.use(
  minifyHTML({
    override: process.env.NODE_ENV === 'production',
    exception_url: false,
    htmlMinifier: {
      removeComments: true,
      collapseWhitespace: true,
      collapseBooleanAttributes: true,
      removeAttributeQuotes: true,
      removeEmptyAttributes: true,
      minifyJS: true,
      minifyCSS: true,
    },
  }),
)

if (process.env.NODE_ENV === 'production') {
  app.set('trust proxy', 1)
}

// view engine setup
app.set('views', path.join(__dirname, 'views'))
app.engine('.hbs', hbs.engine)
app.set('view engine', '.hbs')

app.use(logger('dev'))
app.use(express.json())
app.use(express.urlencoded({ extended: false }))
app.use(cookieParser())
app.use(express.static(path.join(__dirname, 'public')))

app.use(
  '/internal',
  (req, res, next) => {
    next()
  },
  internalRouter,
)

// create a middleware to capture domain slug
app.use(async (req, res, next) => {
  // get the url
  // get process.env.DOMAIN_QUERY_PARAM
  // get process.env.DEBUG
  // get process.env.ROOT_DOMAIN
  // domain slug is = if its in debug mode, then DOMAIN_QUERY_PARAM key query value from the url
  // else it is <slug>.ROOT_DOMAIN from base url of the request
  // if domain slug is not found, then redirect to 404 page because we only serve with some domain
  // if domain slug is found, then set it in res.locals.domainSlug
  // then call next()

  if (Number(process.env.DEBUG)) {
    res.locals.domainSlug = req.query[process.env.DOMAIN_QUERY_PARAM] || ''
  } else if (req.hostname) {
    // check if the domain is part of root domain or is it a custom domain
    // if it is a custom domain, then set res.locals.customDomain to true
    // else set it to false

    // split the hostname at the first dot
    const dotIndex = req.hostname.indexOf('.')
    const firstPart = req.hostname.slice(0, dotIndex)
    const secondPart = req.hostname.slice(dotIndex + 1)

    if (secondPart === process.env.ROOT_DOMAIN) {
      res.locals.customDomain = false

      res.locals.domainSlug = firstPart

      if (res.locals.domainSlug === 'publish') {
        res.locals.domainSlug = req.query[process.env.DOMAIN_QUERY_PARAM]
      }
    } else {
      res.locals.customDomain = true
      res.locals.domainSlug = req.hostname
    }
  }

  if (!res.locals.domainSlug) {
    res.locals.domainSlug = ''
    next()
  } else {
    next()
  }
})

// Sitemap
app.get('/sitemap.xml', (req, res) => {
  // genSitemap().then((d) => {
  //   res.set("Content-Type", "text/xml");
  //   res.type("application/xml");
  //   res.send(d);
  // });
})

app.use('/', indexRouter)

// // catch 404 and forward to error handler
// app.use(function (req, res, next) {
//   next(createError(404));
// });

// Run a CRON that runs every 6 hours cacheUsersData
cron.schedule('0 */6 * * *', () => {
  // eslint-disable-next-line no-console
  console.log('Running a task every 6 hours')
  cacheUsers()
  cacheNotes()
})
cacheUsers()
cacheNotes()

// error handler
app.use(function (err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message
  res.locals.error = req.app.get('env') === 'development' ? err : {}

  // render the error page
  res.status(err.status || 500)
  // res.render("error");
})

module.exports = app
