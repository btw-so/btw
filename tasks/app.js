var createError = require("http-errors");
var debounce = require("debounce");
require("newrelic");
var express = require("express");
var path = require("path");
var cookieParser = require("cookie-parser");
var cors = require("cors");
var logger = require("morgan");
const { createBullBoard } = require("@bull-board/api");
const { BullAdapter } = require("@bull-board/api/bullAdapter");
const { ExpressAdapter } = require("@bull-board/express");
const { Youtube } = require("@tiptap/extension-youtube");
const { Document } = require("@tiptap/extension-document");
const { Paragraph } = require("@tiptap/extension-paragraph");
const { Image } = require("@tiptap/extension-image");
const { Text } = require("@tiptap/extension-text");
const { Bold } = require("@tiptap/extension-bold");
const { Blockquote } = require("@tiptap/extension-blockquote");
const { OrderedList } = require("@tiptap/extension-ordered-list");
const { BulletList } = require("@tiptap/extension-bullet-list");
const { ListItem } = require("@tiptap/extension-list-item");
const { CharacterCount } = require("@tiptap/extension-character-count");
const { Code } = require("@tiptap/extension-code");
const { CodeBlockLowlight } = require("@tiptap/extension-code-block-lowlight");
const { Dropcursor } = require("@tiptap/extension-dropcursor");
const { Gapcursor } = require("@tiptap/extension-gapcursor");
const { HardBreak } = require("@tiptap/extension-hard-break");
const { Heading } = require("@tiptap/extension-heading");
const { Highlight } = require("@tiptap/extension-highlight");
const { HorizontalRule } = require("@tiptap/extension-horizontal-rule");
const { Italic } = require("@tiptap/extension-italic");
const { Link } = require("@tiptap/extension-link");
const { Mention } = require("@tiptap/extension-mention");
const { Placeholder } = require("@tiptap/extension-placeholder");
const { Strike } = require("@tiptap/extension-strike");
const { TaskItem } = require("@tiptap/extension-task-item");
const { TaskList } = require("@tiptap/extension-task-list");
const { Typography } = require("@tiptap/extension-typography");
const { Underline } = require("@tiptap/extension-underline");
const { Node, mergeAttributes } = require("@tiptap/core");

const Embed = Node.create({
    name: "btw-embed",
    group: "block",
    selectable: true,
    draggable: true,
    atom: true,

    parseHTML() {
        return [
            {
                tag: "btw-embed",
                contentElement: "textarea",
            },
        ];
    },
    addAttributes() {
        return {
            code: {
                default: null,
            },
        };
    },
    renderHTML({ HTMLAttributes }) {
        return ["btw-embed", mergeAttributes(HTMLAttributes)];
    },
    parseHTML() {
        return [
            {
                tag: "btw-embed",
            },
        ];
    },
});

const CustomDocument = Document.extend({
    content: "heading block*",
});
var { Database } = require("@hocuspocus/extension-database");
var { Server } = require("@hocuspocus/server");
var { generateHTML, generateJSON } = require("@tiptap/html");
var { TiptapTransformer } = require("@hocuspocus/transformer");
var extensions = [
    Embed,
    CustomDocument,
    Paragraph,
    Text,
    Bold,
    Blockquote,
    OrderedList,
    BulletList,
    ListItem,
    Code,
    Dropcursor,
    Gapcursor,
    HardBreak,
    Heading,
    Highlight,
    HorizontalRule,
    Italic,
    Link,
    Strike,
    TaskList,
    Typography,
    Underline,
    Image,
    TaskItem.configure({
        nested: true,
    }),
    Placeholder.configure({
        placeholder: ({ node }) => {
            if (node.type.name === "heading") {
                return "Whatâ€™s the title?";
            }

            return "Write something...";
        },
    }),
    Youtube.configure({
        controls: false,
    }),
    CodeBlockLowlight,
    Mention.configure({
        HTMLAttributes: {
            class: "mention",
        },
    }),
];
var MyTipTapTransformer = TiptapTransformer.extensions(extensions);
var MyTipTapTransformerHTML = (json) => generateHTML(json, extensions);
var MyTipTapTransformerJSON = (html) => generateJSON(html, extensions);

var indexRouter = require("./routes/index");
var otpRouter = require("./routes/otp");
var notesRouter = require("./routes/notes");
var listRouter = require("./routes/list");
var filesRouter = require("./routes/files");
var { baseQueue } = require("./services/queue");
var { upsertNote, getNote } = require("./logic/notes");

var {
    getUserFromToken,
    createLoginToken,
    createUser,
    setUserDetails,
} = require("./logic/user");
var db = require("./services/db");

// setting up base user for default mode
createUser({
    email: process.env.ADMIN_EMAIL,
    slug: process.env.ADMIN_SLUG,
});

var app = express();

if (process.env.NODE_ENV == "production") {
    app.set("trust proxy", 1);
}

// view engine setup
app.set("views", path.join(__dirname, "views"));
app.set("view engine", "hbs");

app.use(logger("dev"));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, "public")));

app.use("/", indexRouter);
app.use("/otp", otpRouter);
app.use("/notes", notesRouter);
app.use("/list", listRouter);
app.use("/files", filesRouter);
app.use("/user", require("./routes/user"));

// Queue monitor
const serverAdapter = new ExpressAdapter();
const { addQueue, removeQueue, setQueues, replaceQueues } = createBullBoard({
    queues: [new BullAdapter(baseQueue)],
    serverAdapter: serverAdapter,
});
serverAdapter.setBasePath("/admin/queues");
app.use("/admin/queues", serverAdapter.getRouter());

const companion = require("@uppy/companion");
const UPPY_OPTIONS = {
    filePath: "/",
    server: {
        protocol: !!Number(process.env.DEBUG) ? "http" : "https",
        host: process.env.DOMAIN,
        path: "/companion",
    },
    secret: process.env.SECRET,
    debug: !!Number(process.env.DEBUG),
    providerOptions: {
        s3: {
            key: process.env.S3_KEY,
            secret: process.env.S3_SECRET,
            bucket: process.env.S3_BUCKET,
            endpoint: process.env.S3_ENDPOINT,
            region: "us-east-1",
            acl: process.env.COMPANION_AWS_ACL || "public-read",
            object_url: { public: true },
            getKey: (req, fileName) => {
                return req.query.metadata?.fileName || fileName;
            },
        },
    },
    object_url: { public: true },
    corsOrigins: process.env.COMPANION_CLIENT_ORIGINS,
    uploadUrls: process.env.COMPANION_UPLOAD_URLS,
};
const companionApp = companion.app(UPPY_OPTIONS);

console.log("process.env.COMPANION_AWS_ACL", process.env.COMPANION_AWS_ACL);

app.use(
    "/companion",
    cors({
        origin: process.env.COMPANION_CLIENT_ORIGINS.split(","),
    }),
    companionApp
);

// catch 404 and forward to error handler
app.use(function (req, res, next) {
    next(createError(404));
});

// error handler
app.use(function (err, req, res, next) {
    // set locals, only providing error in development
    res.locals.message = err.message;
    res.locals.error = req.app.get("env") === "development" ? err : {};

    // render the error page
    res.status(err.status || 500);
    res.render("error");
});

let debounced = {};

const yjsServer = Server.configure({
    async onAuthenticate(data) {
        const { token: tokenfingerprint, documentName } = data;

        const token = tokenfingerprint.split(":::")[0];
        const fingerprint = tokenfingerprint.split(":::")[1];

        const user = await getUserFromToken({ token, fingerprint });

        if (!user) {
            throw new Error("Not authorized!");
        }

        const userAccordingToUI = documentName.split("note.")[1].split(".")[0];

        var userFromDB;

        try {
            const note = await getNote({
                id: documentName.split("note.")[1].split(".")[1],
                user_id: user.id,
            });
            if (note) {
                userFromDB = note.user_id;
            }
        } catch (e) {
            // can be a new note. for now supporting new notes direct authentication
            console.log(e);
        }

        // We might have shareable docs later on. For now we only allow the owner
        if (
            (userFromDB && user.id !== userFromDB) ||
            "" + user.id !== userAccordingToUI
        ) {
            throw new Error("Not authorized!");
        }

        // You can set contextual data to use it in other hooks
        return {
            user,
        };
    },
    async onChange(data) {
        const save = () => {
            // Convert the y-doc to something you can actually use in your views.
            // In this example we use the TiptapTransformer to get JSON from the given
            // ydoc.
            const prosemirrorJSON = MyTipTapTransformer.fromYdoc(
                data.document,
                "default"
            );

            var html = MyTipTapTransformerHTML(prosemirrorJSON);

            const user_id = data.context.user.id;
            const id = data.documentName.split("note.")[1].split(".")[1];

            upsertNote({
                id,
                user_id,
                json: prosemirrorJSON,
                html,
            });

            // Maybe you want to store the user who changed the document?
            // Guess what, you have access to your custom context from the
            // onAuthenticate hook here. See authorization & authentication for more
            // details
            // console.log(
            //     `Document ${data.documentName} changed by ${data.context.user.id}`
            // );
        };

        debounced[data.documentName] && debounced[data.documentName].clear();
        debounced[data.documentName] = debounce(() => save(), 4000);
        debounced[data.documentName]();
    },
    port: Number(process.env.YJS_PORT),
    extensions: [
        new Database({
            // Return a Promise to retrieve data â€¦
            fetch: async ({ documentName }) => {
                // console.log("fetching", documentName);
                const id = documentName.split("note.")[1].split(".")[1];
                const user_id = documentName.split("note.")[1].split(".")[0];
                let usecase = documentName.split("note.")[1].split(".").length > 2 ? documentName.split("note.")[1].split(".")[2] : null;


                return new Promise((resolve, reject) => {
                    resolve(
                        db.getTasksDB().then((db) => {
                            return db
                                .query(
                                    `SELECT ydoc from btw.notes where id = $1 and user_id = $2`,
                                    [id, Number(user_id)]
                                )
                                .then(({ rows }) => {
                                    // console.log("Found", rows.length);
                                    if (rows.length > 0) {
                                        if (rows[0].ydoc) {
                                            return rows[0]
                                                ? rows[0].ydoc
                                                : null;
                                        } else if (rows[0] && rows[0].html) {
                                            const json =
                                                MyTipTapTransformerJSON(
                                                    rows[0].html
                                                );
                                            return MyTipTapTransformer.toYdoc(
                                                json,
                                                "default"
                                            );
                                        }
                                    } else {
                                        return null;
                                    }
                                });
                        })
                    );
                });
            },
            // â€¦ and a Promise to store data:
            store: async ({ documentName, state }) => {
                // console.log("storing", documentName);
                const id = documentName.split("note.")[1].split(".")[1];
                const user_id = documentName.split("note.")[1].split(".")[0];
                let usecase = documentName.split("note.")[1].split(".").length > 2 ? documentName.split("note.")[1].split(".")[2] : null;

                return new Promise((resolve, reject) => {
                    resolve(
                        db.getTasksDB().then((db) => {
                            return db.query(
                                `INSERT INTO btw.notes (id, user_id, ydoc, created_at, updated_at, tags) VALUES($1, $2, $3, $4, $5, $6) ON CONFLICT(id, user_id) DO UPDATE SET ydoc = $3, updated_at = CASE WHEN
                                notes.ydoc <> EXCLUDED.ydoc
                                OR FALSE THEN
                                EXCLUDED.updated_at
                                ELSE
                                notes.updated_at
                            END RETURNING ydoc`,
                                [
                                    id,
                                    Number(user_id),
                                    state,
                                    new Date(),
                                    new Date(),
                                    usecase || ""
                                ]
                            );
                        })
                    );
                });
            },
        }),
    ],
});

yjsServer.listen();

module.exports = app;
