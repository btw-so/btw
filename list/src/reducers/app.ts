import { createReducer } from '@reduxjs/toolkit';

import { setAppOptions } from 'actions';

import { AppState } from 'types';

export const appState: AppState = {
  query: 'react',
};

export default {
  app: createReducer<AppState>(appState, builder => {
    builder.addCase(setAppOptions, (draft, { payload }) => {
      draft.query = payload.query;
    });
  }),
};
