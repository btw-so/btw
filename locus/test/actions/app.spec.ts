import { setAppOptions } from 'actions/app';

describe('actions/app', () => {
  describe('setAppOptions', () => {
    it('should return an action', () => {
      expect(setAppOptions({ query: 'react' })).toMatchSnapshot();
    });
  });
});
