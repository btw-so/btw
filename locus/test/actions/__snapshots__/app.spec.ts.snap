// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions/app setAppOptions should return an action 1`] = `
Object {
  "payload": Object {
    "query": "react",
  },
  "type": "SET_APP_OPTIONS",
}
`;
